### PR TITLE
Change password 1st step(sending email with the secret)

### PR DESCRIPTION
**Pull request by vantihovich**
…work within branch

### CODE REVIEW

File: swagger/handlers/handlers.go [hunk 1]
```diff
@@ -5,34 +5,40 @@ import (
...
 type UsersHandler struct {
-	userRepo models.UserRepository
-	cache    Cache
-	jwtParam string
-	cfgLogin cnfg.LoginLimitParameters
+	userRepo   models.UserRepository
+	cache      Cache
+	jwtParam   string
+	cfgLogin   cnfg.LoginLimitParameters
+	mailClient email.MailClient
```
Comment (apaliavy) [inline swagger/handlers/handlers.go:27]: I think `email.Client` might be a better name :)

File: swagger/handlers/handlers.go [hunk 2]
```diff
@@ -363,3 +369,60 @@ func (h *UsersHandler) PasswordReset(w http.ResponseWriter, r *http.Request) {
...
+			log.WithError(err).Info("empty request body")
+			w.WriteHeader(http.StatusBadRequest)
+			return
+		}
+		log.WithError(err).Info("error decoding request body occurred")
+		w.WriteHeader(http.StatusBadRequest)
+		return
+	}
+	//generate random secret to store in DB for recovering Password
+	const letterBytes = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789"
```
Comment (apaliavy) [inline swagger/handlers/handlers.go:394]: I think you can use "golang.org/x/crypto/scrypt" package for salt generation

File: swagger/handlers/handlers.go [hunk 3]
```diff
@@ -363,3 +369,60 @@ func (h *UsersHandler) PasswordReset(w http.ResponseWriter, r *http.Request) {
...
+	rand.Seed(time.Now().UnixNano())
+
+	random := make([]byte, 10)
+	for i := range random {
+		random[i] = letterBytes[rand.Intn(len(letterBytes))]
+	}
+
+	secret := string(random)
```
Comment (apaliavy) [inline swagger/handlers/handlers.go:402]: also please salt generation logic out of handler code

File: swagger/handlers/handlers.go [hunk 4]
```diff
@@ -363,3 +369,60 @@ func (h *UsersHandler) PasswordReset(w http.ResponseWriter, r *http.Request) {
...
+
+	//write the secret to DB
+	exists, err := h.userRepo.WriteSecret(ctx, parameters.Email, secret)
```
Comment (apaliavy) [inline swagger/handlers/handlers.go:405]: WriteSecret isn't obvious, could you please change the method name?

File: swagger/handlers/handlers.go [hunk 5]
```diff
@@ -363,3 +369,60 @@ func (h *UsersHandler) PasswordReset(w http.ResponseWriter, r *http.Request) {
...
+		return
+	}
+	if !exists {
+		log.Info("email does not exist in the system")
+		w.WriteHeader(http.StatusNotFound)
+		w.Write([]byte("Please specify the correct email address"))
+		return
+	}
+	//send email to provided email address
+	err = h.mailClient.SendEmail(parameters.Email, secret)
```
Comment (apaliavy) [inline swagger/handlers/handlers.go:418]: I think you should create a better template for your message :)

File: swagger/handlers/handlers.go (hunk 1 above)
Comment (vantihovich) [reply swagger/handlers/handlers.go:27]: done

File: swagger/handlers/handlers.go (hunk 4 above)
Comment (vantihovich) [reply swagger/handlers/handlers.go:405]: done

File: swagger/handlers/handlers.go (hunk 2 above)
Comment (vantihovich) [reply swagger/handlers/handlers.go:394]: Due to troubles with converting generated by "golang.org/x/crypto/scrypt"  []bytes to string I`ve used "github.com/xyproto/randomstring" instead.

File: swagger/handlers/handlers.go (hunk 3 above)
Comment (vantihovich) [reply swagger/handlers/handlers.go:402]: done

File: swagger/email/email.go [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+			From: &mailjet.RecipientV31{
+				Email: m.senderEmail,
+				Name:  "Messager from login services",
+			},
+			To: &mailjet.RecipientsV31{
+				mailjet.RecipientV31{
+					Email: recipient,
+				},
+			},
+			Subject:  "Your secret key for changing password!",
```
Comment (apaliavy) [inline swagger/email/email.go:35]: You should either pass those messages via method params or create a separate mail functions (like SendForgottenPasswordEmail, SendWelcomeEmail, etc) to make them re-usable

File: swagger/main.go [hunk 7]
```diff
@@ -103,6 +112,10 @@ func service() http.Handler {
 		r.Post("/login", UsersProvider.UserLogin)
 		r.Post("/deactivate", mw.Authorize(cfgJWT.SecretKey, UsersProvider.UserDeactivation))
 		r.Post("/password_reset", mw.Authorize(cfgJWT.SecretKey, UsersProvider.PasswordReset))
+		r.Post("/forgot_password", UsersProvider.ForgotPasswordSendEmail)
+
+		// r.Post("forgot_password_check_secret", UsersProvider.)
```
Comment (apaliavy) [inline swagger/main.go:117]: commented code

File: swagger/email/email.go (hunk 6 above)
Comment (vantihovich) [reply swagger/email/email.go:35]: done

File: swagger/main.go (hunk 7 above)
Comment (vantihovich) [reply swagger/main.go]: removed

**Review by apaliavy — APPROVED** [review]

**Pull request merged**